cadence/_internal/workflow/workflow_engine.py

```diff
@@ -1,23 +1,186 @@
+                
+                # Check if workflow is complete
+                if result is not None:
+                    self._is_workflow_complete = True
+                    logger.info(f"Workflow {self._context.info().workflow_id} completed")
+                
+        except Exception:
+            logger.exception(f"Error executing workflow function for {self._context.info().workflow_id}")
+            raise
+    
+    async def _extract_workflow_input(self, decision_task: PollForDecisionTaskResponse) -> Any:
+        """
+        Extract workflow input from the decision task history.
+        
+        Args:
+            decision_task: The decision task containing workflow history
+            
+        Returns:
+            The workflow input data, or None if not found
+        """
+        if not decision_task.history or not hasattr(decision_task.history, 'events'):
+            logger.warning("No history events found in decision task")
+            return None
+            
+        # Look for WorkflowExecutionStarted event
+        for event in decision_task.history.events:
+            if hasattr(event, 'workflow_execution_started_event_attributes'):
+                started_attrs = event.workflow_execution_started_event_attributes
+                if started_attrs and hasattr(started_attrs, 'input'):
+                    # Deserialize the input using the client's data converter
+                    try:
+                        # Use from_data method with a single type hint of None (no type conversion)
+                        input_data_list = await self._context.client().data_converter.from_data(started_attrs.input, [None])
+                        input_data = input_data_list[0] if input_data_list else None
+                        logger.debug(f"Extracted workflow input: {input_data}")
+                        return input_data
+                    except Exception as e:
+                        logger.warning(f"Failed to deserialize workflow input: {e}")
+                        return None
+        
+        logger.warning("No WorkflowExecutionStarted event found in history")
+        return None
+    
+    def _execute_workflow_function_sync(self, workflow_func: Callable, workflow_input: Any) -> Any:
+        """
+        Execute the workflow function synchronously.
+        
+        Args:
+            workflow_func: The workflow function to execute
+            workflow_input: The input data for the workflow function
+            
+        Returns:
+            The result of the workflow function execution
+        """
+        logger.debug(f"Executing workflow function with input: {workflow_input}")
+        result = workflow_func(workflow_input)
+        
+        # If the workflow function is async, we need to handle it properly
+        if asyncio.iscoroutine(result):
+            # Create a simple event loop for async workflow functions
+            try:
+                loop = asyncio.new_event_loop()
```

Comment:
Should use deterministic event loop instead to avoid I/O interactions. 
```suggestion
                loop = DeterministicEventLoop()
```